### DISCUSSION TITLE
'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited

### DISCUSSION THREAD

**Discussion by mark-barrett** [sqlalchemy/sqlalchemy#9076] [Usage Questions]
We have a legacy task (that we will be moving to Celery soon) that runs using FastAPI Utils events in the same FastAPI worker, every 5 minutes.

It makes various calls to the DB and has been working fine up until recently, until we have started encountering the following:

`'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited`

It's strange because each time it runs, it's almost as if it it's in a different place each time. At one moment it'll give a log like:

```
2023-01-10T13:15:16.474+00:00 | /usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py:1638: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
2023-01-10T13:15:16.474+00:00 | _gen_cache_ok = self.__class__.__dict__.get("inherit_cache", False)
```

```
2023-01-10T13:15:15.017+00:00 | /usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py:1180: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
2023-01-10T13:15:15.017+00:00 | obj.__dict__[self.__name__] = result = self.fget(obj)
```

```
2023-01-10T13:15:11.969+00:00 | /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/loading.py:883: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
2023-01-10T13:15:11.969+00:00 | def _instance(row):
```

```
2023-01-10T13:10:15.978+00:00 | /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/state.py:100: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
2023-01-10T13:10:15.978+00:00 | self.committed_state = {}
```

As you can see, each time that it throws an exception, it's coming from a different place in the library each time. Plus, we aren't calling `db.execute` etc. at all without `await`.

So I'm just kind of confused why it keeps jumping around at each time, if theres anything I can do to get more debugging information from it.

FYI we are using version `1.4.28` and the strange thing is, I cannot seem to replicate this locally.

Any help would be greatly appreciated.

**Comment by zzzeek** [comment]
can you provide a complete stack trace, also what precipitates this event (happens every time you do "X", happens only occasionally)

also what might have happened "recently" :

1. any upgrades in software  / changes in versions?
2. application has acquired much more users / data? (is there a staging version of the app?  no errors?)

what libraries and non-standard APIs are in use (in particular SQLModel, FastAPI specific APIs)

**Reply by mark-barrett** [reply]
I tried to get a trace and I couldn't I think it might be because it's a Warning and not an exception being raised. I enabled tracemalloc though and got two more lines:

```
2023-01-10T16:21:28.588+00:00 /usr/local/lib/python3.9/site-packages/starlette/datastructures.py:522: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
2023-01-10T16:21:28.588+00:00	return self[key]
2023-01-10T16:21:28.588+00:00	Object allocated at (most recent call last):
2023-01-10T16:21:28.588+00:00	File "/usr/local/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", lineno 458
2023-01-10T16:21:28.588+00:00	self._prepare_and_execute(operation, parameters)
```

Whats strange about the above is that the exception isn't even coming from SQLAlchemy anymore directly, it's coming from Starlette.

All the task does is checks for rows in the DB that need to be actioned on and applies business logic to them. For example, there might be a row that indicates a "status" == "pending". The task picks all pending rows up from the DB and for each one, makes a HTTP request to a third party.

recently answers:
1. Thought perhaps we might have changed something, but we haven't changed any libraries in that service from before this was working
2. this error occurs in staging where there are very incoming requests daily

In terms of libraries etc., what I've explained uses only FastAPI, FastAPI utils and SQLAlchemy. No new non-standard APIs have been added.

I may as well mention it as it might be something that is linked to the issue, but one of the models we primarily use in the task got a field added to it that looks like this:

```
linked_id = Column(UUID(as_uuid=True), index=True, nullable=True)
```

Not sure if that has anything to do with it.

**Reply by mark-barrett** [reply]
just as another piece of information, we run uvicorn with multiple workers, I considered that perhaps multiple workers might be the issue (even though we've always had it that way), but even when I reduce that to 1, it's the same issue

**Reply by mark-barrett** [reply]
@zzzeek found the issue. We had changed one of our relationships in our model from:

`parent = relationship("Model", remote_side=[id], lazy="joined", join_depth=1)`

to

`parent = relationship("Model", foreign_keys=[parent_id])`

It made it so that if we were trying to access `model.parent.id` it would throw an exception because, if I'm not mistaken, we can't lazyload like that in asyncio?

**Reply by CaselIT** [reply]
that's correct. What's strange here is that you should get a MissingGreenly error instead of the coroutine not waited error.
Can you post the code snipped of what you were doing with the model that made it error like this?

**Reply by mark-barrett** [reply]
it's funny you say that. For whatever reason, perhaps it's related to the logging configuration in our deployed service, but when I run it locally I do get that error:

```exception.InvalidInitialData: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)```

I did say in the original post that I couldn't replicate it locally, however it turns out I wasn't using the latest version of one of our submodules. When I pulled them it worked.

In terms of the snippet, we were doing this:

`instance.parent.id`

Which obviously because of how the relationships were setup, forced a lazyload to happen.

so a bit of advice to anyone in the future who gets this, it's more than likely SQLAlchemy trying to lazyload something in async when it can't.

**Reply by CaselIT** [reply]
not sure. that exception should happen when executing the `instance.parent.id`. strange that it didn't

**Reply by zzzeek** [reply]
so the "never awaited" warning is likely being shown when an object is being garbage collected, which would explain why the location of the warning is erratic.

**Reply by mildo-crest** [reply]
so, what does this mean? Should we ignore it or not? is there a fix?

**Reply by CaselIT** [reply]
Well no code or example was shared so it's a bit hard to suggest a fix.

In general awaitables should be awaited, but without more information it hard to suggest anything else

**Reply by mildo-crest** [reply]
the warning appear in different tests always pointing to different lines of code in sqlalchemy 
```
/sqlalchemy/engine/result.py:106: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
  indexes = self._indexes_for_keys(keys)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback  
```
```
sqlalchemy/orm/context.py:228: RuntimeWarning: coroutine 'AsyncAdapt_asyncpg_cursor._prepare_and_execute' was never awaited
  return SelectState._column_naming_convention(label_style)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

**Reply by CaselIT** [reply]
seems that an execution is not waited in user code. Without an example of the tests you are running that's all it's possible to say
